docs/python_docs/python/tutorials/getting-started/crash-course/3-autograd.md

```diff
@@ -87,20 +263,278 @@ def f(a):
+below.
+
+```{.python .input  n=18}
+with autograd.record():
+    z = 2*x,2*x*y,2*x+2*y
+    z = np.concatenate([np.expand_dims(zi,axis=0) for zi in z],axis=0)
+print(z.shape)
+z.backward()
+print(np.array_equiv(x.grad.asnumpy(),v.asnumpy()))
+print(np.array_equiv(y.grad.asnumpy(),v1.asnumpy()))
+```
+
+```{.json .output n=18}
+[
+ {
+  "name": "stdout",
+  "output_type": "stream",
+  "text": "(3, 2, 3)\nTrue\nTrue\n"
+ }
+]
+```
+
+Now if we were to sum along the first dimension (which autograd assumes is the
+batch_dimension).
+
+```{.python .input  n=19}
+with autograd.record():
+    z = 2*x,2*x*y,2*x+2*y
+    z = np.sum(np.concatenate([np.expand_dims(zi,axis=0) for zi in z],axis=0),axis=0)
+print(z.shape)
+z.backward()
+print(np.array_equiv(x.grad.asnumpy(),v.asnumpy()))
+print(np.array_equiv(y.grad.asnumpy(),v1.asnumpy()))
+```
+
+```{.json .output n=19}
+[
+ {
+  "name": "stdout",
+  "output_type": "stream",
+  "text": "(2, 3)\nTrue\nTrue\n"
+ }
+]
+```
+
+So why do we automatically `sum`? Well as you can see we could have a problem
+where our shapes wouldn't match. Essentially, if we do the operations all
+seperately then we get 3 different gradients for `x,y`. Which one is correct?
+Therefore, we have to decide how we should combine those operations. This is
+where autograd plays a role by default just summing the arguments. We could just
+perform `mean` if we wanted the average, like so.
+
+```{.python .input  n=20}
+with autograd.record():
+    z = 2*x,2*x*y,2*x+2*y
+    z = np.mean(np.concatenate([np.expand_dims(zi,axis=0) for zi in z],axis=0),axis=0)
+print(z.shape)
+z.backward()
+print(np.array_equiv(x.grad.asnumpy(),v.asnumpy()))
+print(np.array_equiv(y.grad.asnumpy(),v1.asnumpy()))
+v = x.grad.copy()
+v1 = y.grad.copy()
+```
+
+```{.json .output n=20}
+[
+ {
+  "name": "stdout",
+  "output_type": "stream",
+  "text": "(2, 3)\nFalse\nFalse\n"
+ }
+]
+```
+
+Alternatively, we could pass in an array for the gradient with respect to our
+current location "upstream". Please note, the shape must match the output shape.
+
+```{.python .input  n=21}
+with autograd.record():
+    z = 2*x,2*x*y,2*x+2*y
+    z = np.concatenate([np.expand_dims(zi,axis=0) for zi in z],axis=0)
+print(z.shape)
+z.backward(np.full(z.shape,1./3.))
+print(np.array_equiv(x.grad.asnumpy(),v.asnumpy()))
+print(np.array_equiv(y.grad.asnumpy(),v1.asnumpy()))
+```
+
+```{.json .output n=21}
+[
+ {
+  "name": "stdout",
+  "output_type": "stream",
+  "text": "(3, 2, 3)\nTrue\nTrue\n"
+ }
+]
+```
+
+As you can see there are 3 values along the dimension 0, so taking a `mean`
+along this access is the same as summing that axis and multiplying by `1/3`.
+
+## Advanced MXNet ndarray operations with Autograd
+
+Additionally, we can control gradients for different ndarray operations. For
+instance, perhaps I want to check that the gradients are propogating properly?
+
+```{.python .input  n=22}
+with autograd.record():
+    y = 3*x
+    y=y.detach()
+    y.attach_grad()
```

Comment:
attach_grad() implicitly calls detach() function, so we will forget the computation graph up till the point where we attach gradient to the intermediate variable, so we end up losing part of the computation graph. Are you sure things actually works correctly here nevertheless?